### ISSUE TITLE
Use SHVDN from non-script thread?

### ISSUE THREAD

**Issue by tcpie** [crosire/scripthookvdotnet#977]
Sourcemod (server-side modification framework for Source-engine games) provides [events](https://wiki.alliedmods.net/Events_(SourceMod_Scripting)#Hooking_Events), which Sourcemod-plugins can use to be notified of function calls and/or modify such calls. This is a very powerful feature, and I have been playing around with getting something similar to work for GTA:V.

For example: I hooked a function that is called when a Ped is hurt by some Entity. My detour function can access and adjust the function parameters, so the hooking part is working well.*

Now, I would like to access a SHVDN script / call natives from my detour function. I don't know how to do this, as I don't understand the SHVDN architecture well enough. Obviously, simply invoking natives from the detour function (I'm using C++/CLI) doesn't work and leads to crashes. Calling std::thread::this_thread::sleep_for() also suspends the Script thread, which leads me to believe that the Script thread is waiting for the (hooked) game thread, so doing things asynchronously probably won't work either.

I'm now thinking that it may be possible to change the hooked thread's "context" (e.g., AppDomain to interact with SHVDN, and the [script thread context](https://github.com/citizenfx/fivem/blob/master/code/components/rage-scripting-five/include/scrThread.h) to be able to call natives). Would this be a viable option? I don't have a clue how to go about this, your suggestions are welcome!

*: simply blocking the hooked function and later calling this function from within SHVDN isn't possible at the moment, as I haven't been able to properly reverse the function's signature (this is challenging for x64 with the parameters being passed by registers). I'm still able to hook the function (in fact, any arbitrary function), as the hooking trampoline saves all of the registers that can be involved in parameter-passing.

**Comment by tcpie** [comment]
OK, I don't think this is going to work with the way SHVDN is set up at the moment.

As crosire explained [here](https://github.com/crosire/scripthookvdotnet/issues/620#issuecomment-305914509), natives in SHVDN are called by dispatching them to the 'main' script thread.

Though I don't think crosire states the reason for this, I presume this is because GTA:V's natives use thread-local storage (TLS). Thus, calling *some* natives from an "invalid" thread will result in a crash. One way around this issue is, indeed, dispatching native calls to the main thread, and calling natives from there. However, this impacts performance (as has been noted in other issues as well), and makes doing the things I would like to do more difficult (impossible?).

A more clean approach (imho) would be impersonating the GTA:V native script threads. This can be done by copying the TLS pointer, which is kept at [GS+0x58](https://en.wikipedia.org/wiki/Win32_Thread_Information_Block). This is also how other GTA:V scripting frameworks approach this issue (for educational purposes I'm trying to implement this myself, but still facing some issues).

In conclusion: SHVDN's approach works well for simple scripting mods, is not very performant, and doesn't appear well-suited for more advanced use cases (e.g., calling natives from hooked function detours).

**Comment by kagikn** [comment]
I think this issue is worth opening because this issue is related to advanced scripting, so let me reopen this.

As @tcpie said, this hook uses separate threads for scripts and switch to the main thread if needed (e.g. native function calls).
With this approach, the script domain class can stop a script if it does not finish its execution (or call any methods that must be done on the main thread) on each tick for 5 seconds. However, the approach prevents scripts from calling any functions that must be called in the main thread for TLS stuff such as [the function for script GUIDs from entity addresses](https://github.com/crosire/scripthookvdotnet/blob/24666753c9d71e25cad1e408464e120ecc129279/source/core/NativeMemory.cs#L1530) and lose performance due to thread switching.
If @crosire allows SHVDN to lose the ability to stop scripts that is not responded for 5 seconds, we can do some more advanced things in SHVDN, I believe. Btw, [`Thread.Abort()` is **STRONGLY** not recommended to call.](https://docs.microsoft.com/en-us/dotnet/api/system.threading.thread.abort?view=net-5.0).

**Comment by cpast** [comment]
Based on my experience (and I checked the code and think this is correct), it already works fine to hook a native game function and call natives from the callback. I've never encountered an issue with it, and `ScriptDomain` seems to correctly handle `IScriptTask`s that are created on the main thread.

**Comment by Sardelka9515** [comment]
I think the issue is related to SHV. Technically SHVDN supports calling natives from game thread:
https://github.com/crosire/scripthookvdotnet/blob/e0d6d3563562844a470d72dce4af0d2e609aa92a/source/core/ScriptDomain.cs#L542 

As you can see, it doesn't dispatch task if you're running in main thread.

If you're not using game thread nor managed script thread, an exception could be thrown, but won't lead to a crash.(unless you don't handle exception at all)

 if you experience a violent crash, then i think the problem is the way SHV calls native(which SHVDN uses under the hood). When you're calling a native, you first need to invoke `nativeInit`, push arguments, then `nativeCall`, which invokes the actual handler, if you call a new native again in the handler, new `nativeInit` ,`nativeCall` will be called, which will mess up SHV's native call context.


My guess is that your hooked function was probably called from SHV, which will crash the game due to the reason above. I may be wrong, though.

**Comment by kagikn** [comment]
Man, I didn't notice @tcpie mentioned the trick I did with #1172 to improve performance. Now we shouldn't face serious performance issues if we implement enumerators that fetch handles from some entity pool (swap the TLS pointer when they need to create new handle/GUID and revert once done, just like RPH).
